open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/App2AppAuthenticator.java

```diff
@@ -0,0 +1,214 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.wso2.openbanking.accelerator.identity.app2app;
+
+import com.nimbusds.jwt.SignedJWT;
+import com.wso2.openbanking.accelerator.common.exception.OpenBankingException;
+import com.wso2.openbanking.accelerator.common.util.JWTUtils;
+import com.wso2.openbanking.accelerator.identity.app2app.exception.JWTValidationException;
+import com.wso2.openbanking.accelerator.identity.app2app.model.DeviceVerificationToken;
+import com.wso2.openbanking.accelerator.identity.app2app.utils.App2AppAuthUtils;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.application.authentication.framework.AbstractApplicationAuthenticator;
+import org.wso2.carbon.identity.application.authentication.framework.FederatedApplicationAuthenticator;
+import org.wso2.carbon.identity.application.authentication.framework.context.AuthenticationContext;
+import org.wso2.carbon.identity.application.authentication.framework.exception.AuthenticationFailedException;
+import org.wso2.carbon.identity.application.authentication.framework.model.AuthenticatedUser;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.DeviceHandler;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.exception.PushDeviceHandlerClientException;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.exception.PushDeviceHandlerServerException;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.impl.DeviceHandlerImpl;
+import org.wso2.carbon.user.api.UserRealm;
+import org.wso2.carbon.user.api.UserStoreException;
+
+import java.text.ParseException;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * App2App authenticator for authenticating users from native auth attempt.
+ */
+public class App2AppAuthenticator extends AbstractApplicationAuthenticator
+        implements FederatedApplicationAuthenticator {
+
+    private static final Log log = LogFactory.getLog(App2AppAuthenticator.class);
+    private static final long serialVersionUID = -5439464372188473141L;
+    private static DeviceHandler deviceHandler;
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getName() {
+
+        return App2AppAuthenticatorConstants.AUTHENTICATOR_NAME;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getFriendlyName() {
+
+        return App2AppAuthenticatorConstants.AUTHENTICATOR_FRIENDLY_NAME;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
```

Comment:
Shall we explain what happens inside this method?

Reply:
Resolved in [cbf2bbd](https://github.com/wso2/financial-open-banking/pull/67/commits/cbf2bbd7f62fb95d3d0e721f80e52b914d8fe80d).